[SPARK-18569][ML][R] Support RFormula arithmetic, I() and spark functions

ozancicek opened the pull request:
## What changes were proposed in this pull request?

Added support for arithmetic expressions, use of spark functions and registered udf's inside formula so that expressions like this work as intended in `RFormula`;

`log(y) ~ a + pow(b, 2)`  ->  `log()` and `pow()` spark functions are used on `y` and `b`
`I(y +b) ~ a + x` -> The label term is sum of `y` and `b`

Udf's can also be used once they're registered:
```scala
  val registeredUdf = spark.udf.register("plusTwo", (x: Int) => (x + 2))
  val formula = new RFormula()
    .setFormula("plusTwo(y) ~ a + plusTwo(b)")

  val df = Seq((1, 4, 4), (2, 5, 6)).toDF("y", "a", "b")
  val model = formula.fit(df)
  model.transform(df).show()

+---+---+---+---------+-----+
|  y|  a|  b| features|label|
+---+---+---+---------+-----+
|  1|  4|  4|[4.0,6.0]|  3.0|
|  2|  5|  6|[5.0,8.0]|  4.0|
+---+---+---+---------+-----+

```
Summary of changes:
- Added `EvalExprParser` trait for parsing arithmetic expressions inside formula
- Added `ExprSelector` transformer which adds columns to a dataframe using `expr` spark function
- Add `ExprSelector` and `ColumnPruner` stages for parsed arithmetic expressions.

The criteria for parsing a term as arithmetic expression is that it's inside 'I(' expression ')', or it's a function (function criteria is ascii chars + alphanumeric chars + '(' args ')' ). From parsed arithmetic expressions, features are generated using `expr` spark function, so anything that can be used with `expr` function should be valid. Due to many ways `expr` function can be used, whether the expression has a valid syntax or whether the function is defined or not is not checked, only parsing rule is balanced parentheses.

## How was this patch tested?

Unit tests to RFormulaParserSuite and RFormulaSuite

@srowen @felixcheung

ozancicek committed:
[SPARK-18569][ML][R]Support RFormula arithmetic, I() and spark functions

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -126,7 +127,13 @@ private[feature] trait RFormulaBase extends HasFeaturesCol with HasLabelCol with
  * Implements the transforms required for fitting a dataset against an R model formula. Currently
- * we support a limited subset of the R operators, including '~', '.', ':', '+', '-', '*' and '^'.
+ * we support a limited subset of the R operators, including '~', '.', ':', '+', '-', '*', '^'
+ * and 'I()'. Arithmetic expressions which use spark functions or registered udf's are
```
udf's -> UDFs

ozancicek replied:
fixed

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -362,6 +381,16 @@ class RFormulaModel private[feature](
+      val dummyDF = spark.createDataFrame(dummyRDD, schema)
+        .withColumn(resolvedFormula.label, expr(resolvedFormula.label))
+      val nullable = dummyDF.schema(resolvedFormula.label).dataType match {
+        case _: NumericType | BooleanType => false
```
I think it can just be `case NumericType | BooleanType`?

ozancicek replied:
I thought so too! Then I got this error;

```
[error] .../spark/mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:390: pattern type is incompatible with expected type;
[error]  found   : org.apache.spark.sql.types.NumericType.type
[error]  required: org.apache.spark.sql.types.DataType
[error] Note: if you intended to match against the class, try `case _: NumericType`
[error]         case NumericType | BooleanType => false


```

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala:
```diff
@@ -269,40 +310,78 @@ private[ml] object RFormulaParser extends RegexParsers {
 
   private val intercept: Parser[Term] =
-    "([01])".r ^^ { case a => Intercept(a == "1") }
+    "([01])".r.n ^^ { case a => Intercept(a == "1") }
```
Pardon, what does `.n` do?

ozancicek replied:
Just introduced a shorter way of escaping whitespace around the tokens. Uses the implicit class defined inside the object on line 290:

```scala
  private val space = "[ \\n]*".r

  /* Utility function for skipping whitespace around a regex. */
  private implicit class RegexParserUtils(val r: Regex) {
    def n: Parser[String] = (space ~> r <~ space)
  }

```
Given a regexp, returns a string parser which skips the surrounding white space

srowen replied:
Oh I see it now. Hm, I think that will be a little more confusing than it's worth to readers. Just add space to the regexes for clarity, I'd say.

ozancicek replied:
Got it, I replaced it with a normal function call.

srowen reviewed (commented):
I don't know this well enough to really review, but it seems plausible.

ozancicek committed:
udf's -> UDFs

felixcheung commented:
thanks! this could be very useful. please give me a bit of time to come back to review.

ozancicek committed:
replace implicit class method call with explicit function

dongjoon-hyun commented:
ok to test

SparkQA commented:
**[Test build #107310 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107310/testReport)** for PR 24939 at commit [`88be1dc`](https://github.com/apache/spark/commit/88be1dcc67283c90922512477062805f63105703).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12455/
Test PASSed.

SparkQA commented:
**[Test build #107310 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107310/testReport)** for PR 24939 at commit [`88be1dc`](https://github.com/apache/spark/commit/88be1dcc67283c90922512477062805f63105703).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107310/
Test FAILed.

ozancicek committed:
Update ResolvedRFormula reference in pyspark doctest

SparkQA commented:
**[Test build #107342 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107342/testReport)** for PR 24939 at commit [`cdf06bd`](https://github.com/apache/spark/commit/cdf06bd4352914f2745c18f5c30b9b67c8ad15c2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12484/
Test PASSed.

SparkQA commented:
**[Test build #107342 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107342/testReport)** for PR 24939 at commit [`cdf06bd`](https://github.com/apache/spark/commit/cdf06bd4352914f2745c18f5c30b9b67c8ad15c2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107342/
Test PASSed.

felixcheung commented:
can we get more test coverage for this change? https://github.com/apache/spark/pull/24939/commits/88be1dcc67283c90922512477062805f63105703
seems a bit more involved

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -362,6 +381,16 @@ class RFormulaModel private[feature](
+      val dummyRDD = spark.sparkContext.parallelize(Seq(Row.empty))
+      val dummyDF = spark.createDataFrame(dummyRDD, schema)
+        .withColumn(resolvedFormula.label, expr(resolvedFormula.label))
+      val nullable = dummyDF.schema(resolvedFormula.label).dataType match {
```
hmm, I'm not sure about this - it's creating a session (referencing the default anyway) and then making a RDD on an empty Row? is it no other way to do this?

ozancicek replied:
In order to create a schema with arbitrary types, I actually borrowed this pattern from here: https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/SQLTransformer.scala#L79

However, the datatype check to determine whether the field should be nullable seems unnecessary here, because only `NumericType` and `BooleanType` seems to be supported, so I modified here to return schema without checking datatype. https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala#L390

ozancicek replied:
I'm resolving this as this part got removed due to the suggested changes.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -614,3 +652,80 @@ private object VectorAttributeRewriter extends MLReadable[VectorAttributeRewrite
+      DefaultParamsWriter.saveMetadata(instance, path, sc)
+      // Save model data: exprsToSelect
+      val data = Data(instance.exprsToSelect.toSeq)
+      val dataPath = new Path(path, "data").toString
```
something more unique? this could conflict easily as "data"

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -614,3 +652,80 @@ private object VectorAttributeRewriter extends MLReadable[VectorAttributeRewrite
+ * @param exprsToSelect set of string expressions to be added as a column to the dataframe.
+ *                      The name of the columns will be identical to the expression
+ */
+private class ExprSelector(
```
I'm unsure about adding a "hidden" stage and MLReader/MLWriter for this - could you explain more why this is necessary?

ozancicek replied:
As you suspect, having an extra hidden stage isn't really essential here. I only added it to have less coupling between RFormula and RFormulaModel classes. This can be very well done without it. 

Roughly, this is what RFormula and RFormulaModel classes are doing;
```scala

class RFormula(..., formula) {
   def fit(df) = {
     val parsedFormula = parse(formula)
     var stages = ArrayBuffer()

     val featureColumns = parsedFormula.terms.map {
         ...
         stages += OneHotEncoder()
         ...
     }
    stages += VectorAssembler(featureColumns)
    val pipeline = Pipeline(stages)
    RFormulaModel(parsedFormula, pipeline)
  }
}

class RFormulaModel(parsedFormula, pipeline) {
  def transform(df) = {
    val withFeatures = pipeline.transform(df)
    transformLabel(withFeatures)
  }
}
```
In order to assemble arithmetic expressions in a feature column with `VectorAssembler`, the dataframe which is transformed by `RFormulaModel` needs to have these columns. One way would be to simply add these transformations inside `RFormulaModel.transform` method,  or another way would be to add a pipelined stage inside `RFormula.fit` method. Seeing that all feature column related transformations are done at `RFormula.fit` method, I chose to add a pipelined stage inside `RFormula.fit` method. 

But indeed, having a transformer for just executing a couple of `expr` functions could be too much. If you think it's unnecessary to add an extra stage, let me know and I'll move it's transformations to `RFormulaModel` class.

felixcheung replied:
if the pipeline model is persisted the "hidden" stage will show up, I think? it's good intention for the abstract but probably best not to add to the confusion there also since the code isn't too complicated. if sharing the logic is desired later we could refactor it but likely not as a private stage anyway.

ozancicek replied:
Got it, thanks! I removed the hidden stage.

ozancicek committed:
Add whitspace skip tests, unique names for path and remove extra schema check

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12859/
Test PASSed.

SparkQA commented:
**[Test build #107736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107736/testReport)** for PR 24939 at commit [`9b1d1e4`](https://github.com/apache/spark/commit/9b1d1e4f1ded80c18e65eaa422acbf1a356364f6).

SparkQA commented:
**[Test build #107736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107736/testReport)** for PR 24939 at commit [`9b1d1e4`](https://github.com/apache/spark/commit/9b1d1e4f1ded80c18e65eaa422acbf1a356364f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107736/
Test PASSed.

ozancicek committed:
Remove hidden transformation stage, add its transformations to RFormulaModel

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12905/
Test PASSed.

SparkQA commented:
**[Test build #107786 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107786/testReport)** for PR 24939 at commit [`4782ebd`](https://github.com/apache/spark/commit/4782ebd2b09ea256415f6d6b066617e4b128a7cc).

SparkQA commented:
**[Test build #107786 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107786/testReport)** for PR 24939 at commit [`4782ebd`](https://github.com/apache/spark/commit/4782ebd2b09ea256415f6d6b066617e4b128a7cc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107786/
Test PASSed.

ozancicek commented:
> can we get more test coverage for this change? [88be1dc](https://github.com/apache/spark/commit/88be1dcc67283c90922512477062805f63105703)
> seems a bit more involved

Added test cases on [9b1d1e4](https://github.com/apache/spark/commit/9b1d1e4f1ded80c18e65eaa422acbf1a356364f6) to cover the whitespace related changes on [88be1dc](https://github.com/apache/spark/commit/88be1dcc67283c90922512477062805f63105703).

ozancicek commented:
friendly ping @felixcheung

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -404,6 +429,30 @@ class RFormulaModel private[feature](
+
+  private def transformSelectExprsSchema(schema: StructType): StructType = {
+    val spark = SparkSession.builder().getOrCreate()
+    val dummyRDD = spark.sparkContext.parallelize(Seq(Row.empty))
```
should it use http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.SparkContext@emptyRDD[T](implicitevidence$8:scala.reflect.ClassTag[T]):org.apache.spark.rdd.RDD[T]

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -404,6 +429,30 @@ class RFormulaModel private[feature](
   }
 
+  private def foldExprs(dataframe: DataFrame)(f: (DataFrame, String) => DataFrame): DataFrame =
+    resolvedFormula.evalExprs.foldLeft(dataframe)(f)
```
this is very minor - perhaps save `resolvedFormula.evalExprs.foldLeft(dataframe)(f)` in a field, next to resolvedFormula which already is

ozancicek replied:
I couldn't quite understand this one. Did you mean save it as a temporary variable and return that?

felixcheung replied:
I mean instead of calling this as a method twice, save the return into a local variable and re-use it (and maybe we don't need a method then)

ozancicek replied:
Got it now, thanks.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala:
```diff
@@ -247,9 +273,24 @@ private[ml] case class Terms(terms: Seq[Term]) extends Term {
+private[ml] object RFormulaParser extends RegexParsers with EvalExprParser {
+
+  /**
+   * Whether to skip whitespace in literals and regex is currently only achived with
```
`achieved`

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala:
```diff
@@ -247,9 +273,24 @@ private[ml] case class Terms(terms: Seq[Term]) extends Term {
+
+  /**
+   * Whether to skip whitespace in literals and regex is currently only achived with
+   * a global switch, and by default it's skipped. We'd like it to be skipped for most parsers,
```
when you refer to a switch.. is it supposed to be configurable?

ozancicek replied:
Oh, when I re-read that comment I realized I didn't make much sense, sorry about that! Let me try to rephrase.

From the perspective of RFormulaParser, it shouldn't be configurable. That switch refers to whitespace skipping behaviour of RegexParsers. RegexParsers don't provide a way to enable-disable whitespace skipping for individual parsers defined inside RegexParsers. There is only a class-wide skipWhitespace flag which disables or enables whitespace skipping for all parsers, and default behaviour is to skip all whitespace. So to be able to parse whitespace for selected parsers, the default behaviour is disabled and whitespace is skipped on the remaining parsers.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala:
```diff
@@ -269,40 +310,78 @@ private[ml] object RFormulaParser extends RegexParsers {
+ * or anything inside `I()`. A valid expression is any string-parentheses product,
+ * such that if there are any parentheses ('(' or ')') they're all balanced.
+ */
+private[ml] trait EvalExprParser extends RegexParsers {
```
is there a reason this isn't doing skipSpace here?

ozancicek replied:
This one is responsible for parsing the function identifier, parenthesis and everything inside the parenthesis. Whatever inside the parenthesis is parsed with whitespaces, so no need to skip there. Function identifiers don't have whitespaces as well. One thing to additionally skip might be the whitespace between the function identifier and the trailing parenthesis, such as;

`y ~ I ( log(b) )`
or
`y ~ exp (b)`

Currently, the above formulas would fail to parse due to the space between the parenthesis and function identifier, e.g `I (` and `exp (`. Should these be allowed? It's not so hard to enable these, but I haven't really given a good thought on what might be unwanted side effects.

ozancicek replied:
It looks like spark sql functions skip the whitespace between function identifier and parenthesis, so I enabled skipping the same whitespace.

felixcheung reviewed (commented):
this is very good and can be super powerful. sorry about the delay but I'm back on and can help to merge this.

a few minor comments, please see if you have some time to address them. thanks for this contribution!

question:
- does this break model persistence? a saved model from spark 2.3 or 2.4 getting loaded in a newer spark, for example?
- generally, I think skipSpace is a good idea, could you add more tests for this since it changes existing behavior and not just for `I()`?

ozancicek commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -216,9 +226,14 @@ class RFormula @Since("1.5.0") (@Since("1.5.0") override val uid: String)
-    // First we index each string column referenced by the input terms.
+    // Add evaluated expressions to the dataset
+    val selectedCols = resolvedFormula.evalExprs
+      .map(col => expr(col).alias(col)) ++ dataset.columns.map(col(_))
```
Most of the time the name stays the same, but there are still some cases name might change. For example when casting is involved;

```scala
import org.apache.spark.sql._
import org.apache.spark.sql.functions._

val spark = SparkSession.builder()
    .master("local")
    .getOrCreate()
import spark.implicits._
Seq(1, 2, 3, 4, 5).toDF.select(expr("log(value)")).show()

+-------------------------------+
|LOG(E(), CAST(value AS DOUBLE))|
+-------------------------------+
|                            0.0|
|             0.6931471805599453|
|             1.0986122886681098|
|             1.3862943611198906|
|             1.6094379124341003|
+-------------------------------+
```

ozancicek commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -126,7 +127,13 @@ private[feature] trait RFormulaBase extends HasFeaturesCol with HasLabelCol with
  * Implements the transforms required for fitting a dataset against an R model formula. Currently
- * we support a limited subset of the R operators, including '~', '.', ':', '+', '-', '*' and '^'.
+ * we support a limited subset of the R operators, including '~', '.', ':', '+', '-', '*', '^'
+ * and 'I()'. Arithmetic expressions which use spark functions or registered UDFs are
```
Since expr function is used, I think it should be Spark SQL functions so I'm changing it as such.

ozancicek committed:
Extend whitespace tests, support loading old models and minor fixes

ozancicek commented on mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala:
```diff
@@ -198,4 +198,31 @@ class RFormulaParserSuite extends SparkFunSuite {
+  test("parse functions") {
+    checkParse("y ~ I(a+b) + c", "y", Seq("a+b", "c"))
+    checkParse("y ~ I(a+b)*c", "y", Seq("a+b", "c", "a+b:c"))
+    checkParse("y ~ (I((a+b)) + c)^2", "y", Seq("(a+b)", "c", "(a+b):c"))
```
Added additional tests with whitespace

ozancicek commented on mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala:
```diff
@@ -198,4 +198,31 @@ class RFormulaParserSuite extends SparkFunSuite {
+    checkParse("y ~ I(log(a)*(log(a)*2)) + b", "y", Seq("log(a)*(log(a)*2)", "b"))
+    checkParse("y ~ exp(a) + (b + c)", "y", Seq("exp(a)", "b", "c"))
+    checkParse("log(y) ~ a + log(b)", "log(y)", Seq("a", "log(b)"))
+    checkParse("I(c+d) ~ a + log(b)", "c+d", Seq("a", "log(b)"))
```
also added tests for those functions

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15135/
Test PASSed.

SparkQA commented:
**[Test build #110127 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110127/testReport)** for PR 24939 at commit [`a11e75b`](https://github.com/apache/spark/commit/a11e75baa7ed1022c40f9597196385517e017790).

ozancicek commented:
> this is very good and can be super powerful. sorry about the delay but I'm back on and can help to merge this.
> 
> a few minor comments, please see if you have some time to address them. thanks for this contribution!
> 
> question:
> 
> * does this break model persistence? a saved model from spark 2.3 or 2.4 getting loaded in a newer spark, for example?
> * generally, I think skipSpace is a good idea, could you add more tests for this since it changes existing behavior and not just for `I()`?

Thanks for the review and updates!

I haven't initially thought about persistence compatibility, thanks for the notice. As there is only one additional field to be persisted, I've updated the model loader not to break when loading an old model which doesn't include that additional field. I also added additional tests for skipSpace to RFormulaParserSuite. If there is some additional tests you have in mind, please let me know and I'll add more. All the changes are in the most recent commit.

SparkQA commented:
**[Test build #110127 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110127/testReport)** for PR 24939 at commit [`a11e75b`](https://github.com/apache/spark/commit/a11e75baa7ed1022c40f9597196385517e017790).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110127/
Test PASSed.

felixcheung commented:
@srowen do you know what's our Spark ML policy on persisted model forward/backward compatibility?

srowen commented:
It should be possible to read older models in newer versions, but not necessary to ensure that newer models are still readable in older versions -- although that's a nice to have if possible.

ozancicek committed:
Save foldLeft result in a field and reuse in transform method

SparkQA commented:
**[Test build #110348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110348/testReport)** for PR 24939 at commit [`ed83506`](https://github.com/apache/spark/commit/ed8350661601bb7feeb9cd862a81e79cf5ee066a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15338/
Test PASSed.

SparkQA commented:
**[Test build #110348 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110348/testReport)** for PR 24939 at commit [`ed83506`](https://github.com/apache/spark/commit/ed8350661601bb7feeb9cd862a81e79cf5ee066a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110348/
Test PASSed.

ozancicek committed:
skip whitespace between function name and open paren

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15431/
Test PASSed.

SparkQA commented:
**[Test build #110444 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110444/testReport)** for PR 24939 at commit [`56e36f1`](https://github.com/apache/spark/commit/56e36f1b269e26e04005ee79570558b2f8146d1f).

SparkQA commented:
**[Test build #110444 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110444/testReport)** for PR 24939 at commit [`56e36f1`](https://github.com/apache/spark/commit/56e36f1b269e26e04005ee79570558b2f8146d1f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110444/
Test PASSed.

HyukjinKwon commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15842/
Test PASSed.

SparkQA commented:
**[Test build #110680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110680/testReport)** for PR 24939 at commit [`56e36f1`](https://github.com/apache/spark/commit/56e36f1b269e26e04005ee79570558b2f8146d1f).

SparkQA commented:
**[Test build #110680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110680/testReport)** for PR 24939 at commit [`56e36f1`](https://github.com/apache/spark/commit/56e36f1b269e26e04005ee79570558b2f8146d1f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110680/
Test FAILed.

ozancicek commented:
> **[Test build #110680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110680/testReport)** for PR 24939 at commit [`56e36f1`](https://github.com/apache/spark/commit/56e36f1b269e26e04005ee79570558b2f8146d1f).
> 
> * This patch **fails PySpark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

```
FAIL: test_train_prediction (pyspark.mllib.tests.test_streaming_algorithms.StreamingLinearRegressionWithTests)
Test that error on test data improves as model is trained.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/mllib/tests/test_streaming_algorithms.py", line 503, in test_train_prediction
    self._eventually(condition)
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/mllib/tests/test_streaming_algorithms.py", line 69, in _eventually
    lastValue = condition()
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/mllib/tests/test_streaming_algorithms.py", line 498, in condition
    self.assertGreater(errors[1] - errors[-1], 2)
AssertionError: 1.672640157855923 not greater than 2
```

I wasn't able to reproduce this failure locally. Could it be a transient error?

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #110944 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110944/testReport)** for PR 24939 at commit [`56e36f1`](https://github.com/apache/spark/commit/56e36f1b269e26e04005ee79570558b2f8146d1f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16053/
Test PASSed.

SparkQA commented:
**[Test build #110944 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110944/testReport)** for PR 24939 at commit [`56e36f1`](https://github.com/apache/spark/commit/56e36f1b269e26e04005ee79570558b2f8146d1f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110944/
Test PASSed.

SparkQA commented:
**[Test build #111690 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111690/testReport)** for PR 24939 at commit [`0455437`](https://github.com/apache/spark/commit/045543774a400fdfba4a907c1310216462eac2cf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16683/
Test PASSed.

ozancicek commented:
> It should be possible to read older models in newer versions, but not necessary to ensure that newer models are still readable in older versions -- although that's a nice to have if possible.

@felixcheung  Regarding changes to persistence, older models are possible to read in newer versions. Newer models should also be possible to read in older versions, the additional column to the persisted case class is added to the end without any changes in column ordering & model is saved as parquet. I've also extended the tests regarding whitespace. Let me know if you have anything additional in mind.

SparkQA commented:
**[Test build #111690 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111690/testReport)** for PR 24939 at commit [`0455437`](https://github.com/apache/spark/commit/045543774a400fdfba4a907c1310216462eac2cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111690/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17261/
Test PASSed.

SparkQA commented:
**[Test build #112278 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112278/testReport)** for PR 24939 at commit [`b7ef3cd`](https://github.com/apache/spark/commit/b7ef3cd9d040048591487a8c3a934b5037956a9f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112278/
Test PASSed.

ozancicek commented:
@zhengruifeng @mengxr is it possible for you to review this PR? @felixcheung was reviewing it but he's been away recently.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -442,11 +487,16 @@ object RFormulaModel extends MLReadable[RFormulaModel] {
       val terms = data.getAs[Seq[Seq[String]]](1)
       val hasIntercept = data.getBoolean(2)
-      val resolvedRFormula = ResolvedRFormula(label, terms, hasIntercept)
+      // Check for old saved models which do not have evalExprs.
```
Here I suggest use method `VersionUtils.majorMinorVersion`

```scala
val (major, minor) = VersionUtils.majorMinorVersion(metadata.sparkVersion)
if (major.toInt < 3) {
    //for old models
} else {
    //for new models
}

```

zhengruifeng reviewed (commented):
I think it is super useful for end users familiar with RFormula. However, I do not know much about it, so mayebe we need to wait @felixcheung

ozancicek committed:
check spark version explicitly when loading model

SparkQA commented:
**[Test build #114533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114533/testReport)** for PR 24939 at commit [`85cf99f`](https://github.com/apache/spark/commit/85cf99fd6d7362cac908ff8208044dc44a585495).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19362/
Test PASSed.

SparkQA commented:
**[Test build #114533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114533/testReport)** for PR 24939 at commit [`85cf99f`](https://github.com/apache/spark/commit/85cf99fd6d7362cac908ff8208044dc44a585495).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114533/
Test PASSed.

viirya reviewed (commented):
Looks useful feature.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -442,11 +488,17 @@ object RFormulaModel extends MLReadable[RFormulaModel] {
+      // Check for old saved models (spark version < 3), they do not have evalExprs.
+      val (major, minor) = VersionUtils.majorMinorVersion(metadata.sparkVersion)
+      val evalExprs = if (major.toInt < 3) Seq[String]() else data.getAs[Seq[String]](3)
+      val resolvedRFormula = ResolvedRFormula(label, terms, hasIntercept, evalExprs)
```
Looks correct. We better have a test for this. We can load a model from 2.4 in a test case. There are few examples in ml.

ozancicek replied:
Thanks, added a testcase for loading a model from 2.4.4.

ozancicek committed:
Add testcase for loading 2.4.4 model, use withColumns & drop

ozancicek commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -344,29 +360,28 @@ class RFormulaModel private[feature](
-    transformLabel(pipelineModel.transform(dataset))
+    val withExprs = foldExprs(dataset.toDF) {
+      case(df, colname) => df.withColumn(colname, expr(colname))
+    }
```
Thanks, I wasn't aware of that function. Switched to withColumns.

ozancicek commented on mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala:
```diff
@@ -344,29 +360,28 @@ class RFormulaModel private[feature](
+    val withFeatures = pipelineModel.transform(withExprs)
+    val withLabel = transformLabel(withFeatures)
+    foldExprs(withLabel) {
+      case(df, colname) => df.drop(col(s"`$colname`"))
```
Indeed, foldLeft is unnecessary. Switched to ```drop(colNames: String*)```

SparkQA commented:
**[Test build #116525 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116525/testReport)** for PR 24939 at commit [`c198669`](https://github.com/apache/spark/commit/c198669123c05e2a195422147a3a22ac4960ea7e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21311/
Test PASSed.

SparkQA commented:
**[Test build #116525 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116525/testReport)** for PR 24939 at commit [`c198669`](https://github.com/apache/spark/commit/c198669123c05e2a195422147a3a22ac4960ea7e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116525/
Test PASSed.

ozancicek committed:
add comment to rformula old model read test

SparkQA commented:
**[Test build #116729 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116729/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21504/
Test PASSed.

SparkQA commented:
**[Test build #116729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116729/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116729/
Test PASSed.

ozancicek commented:
friendly ping @viirya

AmplabJenkins commented:
Can one of the admins verify this patch?

viirya commented:
retest this please

SparkQA commented:
**[Test build #120014 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120014/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24732/
Test PASSed.

SparkQA commented:
**[Test build #120014 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120014/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120014/
Test PASSed.

HyukjinKwon commented:
cc @mengxr and @ericl. cc @shivaram too.

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #124615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124615/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/29236/
Test PASSed.

SparkQA commented:
**[Test build #124615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124615/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124615/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #124676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124676/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/29292/
Test PASSed.

SparkQA commented:
**[Test build #124676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124676/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124676/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #124775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124775/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/29387/
Test PASSed.

SparkQA commented:
**[Test build #124775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124775/testReport)** for PR 24939 at commit [`c3d97a3`](https://github.com/apache/spark/commit/c3d97a351a04474a260ff6ae40909f366a8398cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124775/
Test PASSed.
